app/models/pubmed_source_record.rb

```diff
@@ -79,258 +79,8 @@ def pubmed_update
-    pubmed_record.xpath('MedlineCitation/Article/Abstract/AbstractText').text
+  def source_as_hash
+    Pubmed::MapPubHash.map(source_data)
   end
 
-  def extract_mesh_headings_from_pubmed_record(pubmed_record)
-    mesh_headings_for_record = []
-    pubmed_record.xpath('MedlineCitation/MeshHeadingList/MeshHeading').each do |mesh_heading|
-      descriptors = []
-      qualifiers = []
-      mesh_heading.xpath('DescriptorName').each do |descriptor_name|
-        descriptors << { major: descriptor_name.attr('MajorTopicYN'), name: descriptor_name.text }
-      end
-      mesh_heading.xpath('QualifierName').each do |qualifier_name|
-        qualifiers << { major: qualifier_name.attr('MajorTopicYN'), name: qualifier_name.text }
-      end
-      mesh_headings_for_record << { descriptor: descriptors, qualifier: qualifiers }
-    end
-    mesh_headings_for_record
-  end
-
-  # see https://dtd.nlm.nih.gov/ncbi/pubmed/doc/out/180101/el-Year.html
-  def extract_year_from_pubmed_record(publication)
-    # look for a year in all of the xpath locations in order, pick the first that produces something that looks like a four digit year
-    pubmed_date_xpaths('Year')
-      .map { |path| publication.xpath(path).text.match(/[12][0-9]{3}/).to_s }
-      .compact_blank.first
-  end
-
-  # see https://dtd.nlm.nih.gov/ncbi/pubmed/doc/out/180101/el-Month.html
-  def extract_month_from_pubmed_record(publication)
-    # look for a month in all of the xpath locations in order
-    #  pick the first that produces something that looks like a month (12 or Aug)
-    pubmed_date_xpaths('Month').map do |path|
-      month_value = publication.xpath(path).text
-      match = month_value.match(/\A[012]?[0-9]{1}\z/) # e.g. 01, 11, 06, 6, 12
-      name_match = month_value.match(/\A[a-zA-Z]{3}\z/) # e.g. Aug, aug, Oct
-      match = Date::ABBR_MONTHNAMES.index(name_match.to_s) if name_match # convert month name to number
-      match.to_s
-    end.compact_blank.first
-  end
-
-  # see https://dtd.nlm.nih.gov/ncbi/pubmed/doc/out/180101/el-Day.html
-  def extract_day_from_pubmed_record(publication)
-    # look for a day in all of the xpath locations in order
-    #  pick the first that produces something that looks like a day
-    pubmed_date_xpaths('Day').map do |path|
-      day_value = publication.xpath(path).text
-      match = day_value.match(/\A[0123]?[0-9]{1}\z/) # e.g. 01, 11, 06, 6, 12, 23, 30
-      match.to_s
-    end.compact_blank.first
-  end
-
-  # Convert MEDLINE®PubMed® XML to pub_hash
-  # @param [Nokogiri::XML::Node] publication
-  # @return [Hash<Symbol => Object>] pub_hash
-  # @see https://www.nlm.nih.gov/bsd/licensee/elements_descriptions.html XML Element Descriptions and their Attributes
-  # rubocop:disable Metrics/AbcSize
-  # rubocop:disable Metrics/CyclomaticComplexity
-  # rubocop:disable Metrics/PerceivedComplexity
-  def source_as_hash(publication = Nokogiri::XML(source_data).xpath('//PubmedArticle'))
-    record_as_hash = {}
```

Comment:
We remove all of the pub hash generation related methods to a new class, and replace this `source_as_hash` method with a simple call to our new class when we ask for the pub hash (as shown above)